flang/lib/Optimizer/Builder/FIRBuilder.cpp

```diff
@@ -1541,21 +1541,45 @@ mlir::Value fir::factory::genMaxWithZero(fir::FirOpBuilder &builder,
                                                zero);
 }
 
+static std::pair<mlir::Value, mlir::Type>
+genCPtrOrCFunptrFieldIndex(fir::FirOpBuilder &builder, mlir::Location loc,
+                           mlir::Type cptrTy) {
+  assert(mlir::isa<fir::RecordType>(cptrTy));
+  auto recTy = mlir::dyn_cast<fir::RecordType>(cptrTy);
```

Comment:
nit: assert + dyn_cast can just be `mlir::cast`. See https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates